src/redux/reducers/loading.js

```diff
@@ -0,0 +1,61 @@
+import { actionTypes as dataSetActionTypes } from '../actions/dataSet';
+import { actionTypes as htmlReportActionTypes } from '../actions/htmlReport';
+import { ACTION_TYPES as organisationUnitsActionTypes } from '../actions/organisationUnits';
+import { ACTION_TYPES as reportPeriodActionTypes } from '../actions/reportPeriod';
+import { actionTypes as standardReportActionTypes } from '../actions/standardReport';
+
+export const defaultState = {
+    loading: false,
+    error: '',
+}
+
+export const loading = (state = defaultState, { type, payload } = {}) => {
+    switch (type) {
+        case dataSetActionTypes.LOADING_DATA_SET_OPTIONS_START:
+        case dataSetActionTypes.LOADING_DIMENSIONS_START:
+        case htmlReportActionTypes.LOADING_HTML_REPORT_START:
+        case htmlReportActionTypes.SHARING_DATA_SET_REPORT_COMMENT_START:
+        case organisationUnitsActionTypes.ORGANISATION_UNITS_LOADING_START:
+        case organisationUnitsActionTypes.LOADING_GROUP_SETS_START:
+        case reportPeriodActionTypes.REPORT_PERIOD_TYPES_LOADING_START:
+        case standardReportActionTypes.LOADING_STANDARD_REPORTS_START:
+        case standardReportActionTypes.DELETE_STANDARD_REPORT_START:
+            return {
+                ...state,
+                loading: true,
+                error: '',
+            }
+
+        case dataSetActionTypes.LOADING_DATA_SET_OPTIONS_SUCCESS:
+        case dataSetActionTypes.LOADING_DIMENSIONS_SUCCESS:
+        case htmlReportActionTypes.LOADING_HTML_REPORT_SUCCESS:
+        case htmlReportActionTypes.SHARING_DATA_SET_REPORT_COMMENT_SUCCESS:
+        case organisationUnitsActionTypes.ORGANISATION_UNITS_RECEIVED:
+        case organisationUnitsActionTypes.LOADING_GROUP_SETS_SUCCESS:
+        case reportPeriodActionTypes.REPORT_PERIOD_TYPES_RECEIVED:
+        case standardReportActionTypes.LOADING_STANDARD_REPORTS_SUCCESS:
+        case standardReportActionTypes.DELETE_STANDARD_REPORT_SUCCESS:
+            return {
+                ...state,
+                loading: false,
+            }
+
+        case dataSetActionTypes.LOADING_DATA_SET_OPTIONS_ERROR:
+        case dataSetActionTypes.LOADING_DIMENSIONS_ERROR:
+        case htmlReportActionTypes.LOADING_HTML_REPORT_ERROR:
+        case htmlReportActionTypes.SHARING_DATA_SET_REPORT_COMMENT_ERROR:
+        case organisationUnitsActionTypes.ORGANISATION_UNITS_ERRORED:
+        case organisationUnitsActionTypes.LOADING_GROUP_SETS_ERROR:
+        case reportPeriodActionTypes.REPORT_PERIOD_TYPES_ERRORED:
+        case standardReportActionTypes.LOADING_STANDARD_REPORTS_ERROR:
+        case standardReportActionTypes.DELETE_STANDARD_REPORT_ERROR:
+            return {
+                ...state,
+                loading: false,
+                error: payload,
+            }
+
+        default:
+            return state;
+    }
+}
```

Comment:
Aha, this is where `loading.loading` originates from.... As I said in my previous comment, I think we should distinguish between "fetch-once" and "fetch-repeatably" type of resources here.

Entities like standard-reports and resources, etc. have paging/filtering/deleting, so these actually need a loading state. My personal preference would be to integrate this into the existing reducers for that entity instead of having a generic loading state.

Other entities, like org-units and period-types, are more or less static resources (from the perspective of the reports-app). After they have been fetched nothing happens with them. IMO these don't need a loading state at all. So far all components (I've made) handle the scenario of not loading themselves.

So, if it were up to me I'd like to get rid of this reducer all together.

Reply:
I'll reply here to all the loading state comments you added.

# Why outsourcing the loading state
I outsourced the htmlReport into its own reducer/actions as it's basically reused across the application sections. At first there was a loading state inside this reducer as well, but all section kind of depend on whether something is loading and interaction should be "disabled" when loading to prevent weird behavior (i. e. by showing the loading spinner of the Feedback component). This is also the case for the standard report section, so the `disabled` state for interaction would derive from multiple loading sources which, in my opinion, is an unnecessary overhead if we can simply have one loading that that needs to be set to true/false for the whole application.

# Why having a loading state for the data that's being loaded during app initialization
I think that preventing interaction is also good practice for loading organisation units, data sets in the beginning as well. Why should the user be able to interact with the application while the required data for interaction is being loaded (especially for users with a low-bandwidth internet connection)? At least this was my assumption.

If you agree with this, then checking whether the user can interact or not would derive from "many" loading states of several reducers, which would make the app more complicated than it needs to be.

# Summary of both previous comment section
If preventing the user from interaction with the application is a good idea, and you agree that this is also the case when loading organisation units, data sets, etc., in the beginning, then the loading reducer is a nice way for react to many actions that trigger the loading state to change rather than to spread the loading state across multiple reducers, which results in a more complex and complicated state.

# Why `loading.loading`
The reducer is called loading, that's explains the first `loading`. As requesting an ajax resource can result in an error, I added the property `loadingError` to the state as well. So I wasn't sure how to name the property for the actual `loading` state, which is why there's the name duplication.
We can change this though, I'm not so sure about this anymore as we're storing loading errors in the feedback state as well. Would you agree to change the state from `{ loading: boolean; loadingError: string; }` to `boolean`? Then we'd get rid of the `state.loading.loading`, it'd change to `state.loading`. The loadingError property isn't used anywhere anyways..

Reply:
I've numbered your headers cause I'm lazy:
**reg. 1 and 4:** I don't necessarily disagree with having a global loading state. It might be a good idea for app/section initialisation. However, I still think we need to distinguish between two scenarios:
a. Entities that are being fetched once (in the app or in a section), and are required for the app (or the app-section) to function. During loading we should disable interaction, so these are related to the central app loading state. OrgUnit, PeriodType, etc. would fall under this.
b. Entities that are fed to list components, such as resource,  shouldn't block the interaction with the app in general, rather, it should just show a spinner on the list component. So in my mind, this would NOT be part of the central loading state.

**reg. 3:** OK, you've won me over: having a central loading state like you have implemented is good. But the fact still remains that the stuff that is being fetched once from `App.js` doesn't need a `LOADING_START` type of action. Initially loading state is false and once the response comes it, it can be set to true.

**reg 4:** It would be nice if loading was just a boolean. Otherwise, we could call the reducer `status` and keep the structure as is. `status.loading` `status.error` that makes sense to me.

**In general**:
I don't think the current implementation of the reducer is logically sound. In theory, the state could get toggled between loading true || false very often, which would then trigger a lot of re-renders. To make it more reliable in the current implementation, we would need some sort of lookup object/Map. The keys would be the action types minus `_START` or `_SUCCESS`, and loading should only be toggled to false of all keys are false.

**An alternative**:
Suppose we ditch the general loading state reducer, and instead we introduce a selector. This selector could be used in the `mapStateToProps` function of the app component, and would read the loading state of separate entities and return a boolean. I think I would actually prefer this approach, because the general loading state is actually a derived state, and this is what selectors are for.....

Reply:
I'll summarize a few things first, then comment on your proposals.

### 1. We need a global loading state for blocking interaction for some requests
These will be org units, period types, pagination (standard report and resource section) and html report requests.

### 2. We have some dropdowns which options will be loaded which shouldn't block app interaction.
These loading states should disable interaction with the dropdown the options are loaded for.

### 3. The loading state for values that are being fetched once and during startup of the app are set to true initially, so we don't need to dispatch loading actions for those.

So regarding the idea to use a selector, I think it's a nice idea. This means:
1. We get rid of the loading reducer
1. All reducers that contain state that must be loaded will have a loading state
1. We have a selector for `blockUi`, which is derived from all the ui blocking loading states
1. We get rid of the loadingError state as that's being handled by the feedback component
1. We display the loading state of dropdown components that don't need to block the whole UI on the dropdowns themselves

I have one remark regarding the last one: 
The DropDown component doesn't have a 'loading' property. So I suggest we disable the dropdown and show 'Options are being loaded' as hint text. Do you agree to this?

Reply:
Oh, one thing I just realised: we have the `feedback` component which can also show a loader. I suppose this component should be subscribing to the derived loading state, so it can show the loader.... Not the app component as I suggested earlier.
Then it would be something like this in `src/components/Feedback.js` line 50:
```javascript
showLoader: state.feedback.showLoader || blockUI(state),
```

Reply:
yes, that was my plan anyways :)

Reply:
@HendrikThePendric 
I just noticed that it doesn't make much sense to combine the loading and snackbar components.

What I'll do is:
1. Split those components (but keep the feedback state in one reducer)
1. Put the loading component into the App.js (this way we don't have to calculate the showLoader prop for every component)
1. Put the snackbar component into every section (as the standard report and resource section need to add the onActionClick function to the conf)